### PR TITLE
Postgresql schema upgrade step

### PR DESCRIPTION
**Pull request by UnderGreen**

### CODE REVIEW

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 1]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
+
+		res, err = u.crdClient.MarkPhase(res, phase, crd.StolonUpgradeStatusCompleted)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	return res, nil
+}
+
+func (u *upgradeControl) backupPostgres() error {
```
Comment (knisbet) [inline internal/stolonctl/pkg/cluster/upgrade.go]: One item I'm a little curious about. I didn't track this all the way through, but it looks like the default for BackupPath is /stolon/backup.sql, but I think the container mounts /stolon-data. So I would double check that the backup is being stored persistently. 

Also, I suspect restore could be a bit tricky, if the pod get's scheduled to a node that isn't the one that took the backup.

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 2]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
+					}
+				}
+			}
 		}
-		if err = backupPostgres(config); err != nil {
-			return trace.Wrap(err)
+
+		log.Infof("master: %s, standbys: %v\n", masterNodeName, standbyNodeNames)
+		if masterNodeName == "" {
+			return trace.BadParameter("IP address of keeper master pod is empty")
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/upgrade.go]: How can this be? Maybe rephrase the message to say that the master Pod was not found?

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 3]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
 		}
-		res, err = crdclient.MarkPhase(res, crd.StolonUpgradePhaseBackupPostgres, crd.StolonUpgradeStatusCompleted)
+		err := upgradeControl.upgradePostgresSchema(ctx, pgMaster, masterNodeName, "")
 		if err != nil {
 			return trace.Wrap(err)
 		}
+		for id, standbyNodeName := range standbyNodeNames {
+			err = upgradeControl.upgradePostgresSchema(ctx, pgStandby, standbyNodeName, id)
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/upgrade.go]: What would happen if the schema would fail to update on one of the nodes? How can this be remedied? Do you need to roll anything back in this case?

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 4]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
+		return trace.Wrap(err)
 	}
 	return nil
 }
 
-func backupPostgres(config Config) error {
-	if err := createPgPassFile(config); err != nil {
+func (u *upgradeControl) executePhase(res *crd.StolonUpgradeResource, phase string, fn func() error) (*crd.StolonUpgradeResource, error) {
+	var err error
+	if !u.crdClient.IsPhaseCompleted(res, phase) {
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/upgrade.go]: Invert the condition to avoid indenting:
```go
if u.crdClient.IsPhaseCompleted(res, phase) {
	return nil
}
// handle the case with incomplete phase
```

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 5]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	return res, nil
+}
+
+func (u *upgradeControl) checkStatus(res *crd.StolonUpgradeResource, phase string) (*crd.StolonUpgradeResource, error) {
+	var err error
+	if !u.crdClient.IsPhaseCompleted(res, phase) {
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/upgrade.go]: Invert the condition to avoid indentation.

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 6]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
+
+// deleteDeployment deletes sentinels deployment
+func (u *upgradeControl) deleteDeployment(name string) error {
+	dsClient := u.kubeClient.Client.AppsV1().Deployments(u.config.Namespace)
+	deletePolicy := metav1.DeletePropagationForeground
+	err := dsClient.Delete(name, &metav1.DeleteOptions{
+		PropagationPolicy: &deletePolicy,
+	})
+	if err != nil {
+		return trace.Wrap(err, "cannot delete stolon-sentinel deployment")
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/upgrade.go]: ~~cannot~~ failed to delete
Also consider using `rigging.ConvertErrorWithContext` to translate k8s errors.

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 7]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
+
+// deleteDaemonset deletes keepers daemonset
+func (u *upgradeControl) deleteDaemonset(name string) error {
+	dsClient := u.kubeClient.Client.ExtensionsV1beta1().DaemonSets(u.config.Namespace)
+	deletePolicy := metav1.DeletePropagationForeground
+	err := dsClient.Delete(name, &metav1.DeleteOptions{
+		PropagationPolicy: &deletePolicy,
+	})
+	if err != nil {
+		return trace.Wrap(err, "cannot delete stolon-keeper daemonset")
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/upgrade.go]: Same here about `rigging.ConvertErrorWithContext`

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 8]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
+	}
+	return nil
+}
+
+func (u *upgradeControl) upgradePostgresSchema(ctx context.Context, pgRole, nodeName, id string) error {
+	jobControl, err := rigging.NewJobControl(rigging.JobConfig{u.generateJob(pgRole, nodeName, id), u.kubeClient.Client})
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/upgrade.go]: nit: this could use a bit of formatting for readability:
```go
jobControl, err := rigging.NewJobControl(
	rigging.JobConfig{
		Job:       u.generateJob(pgRole, nodeName, id),
		Clientset: u.kubeClient.Client,
	},
)
```

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 9]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := jobControl.Upsert(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	// wait 5 minutes for completed status of Job
+	return utils.Retry(ctx, 300, time.Second, func() error {
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/upgrade.go]: move the timeout to consts

File: internal/stolonctl/pkg/cluster/upgrade.go (hunk 1 above)
Comment (a-palchikov) [reply internal/stolonctl/pkg/cluster/upgrade.go]: Oh, that's a good catch.

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 10]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Create job to upgrade PostgreSQL schema on keeper master
+	res, err = upgradeControl.executePhase(res, crd.StolonUpgradePhaseUpgradePostgresSchema, func() error {
+		var masterNodeName string
+		standbyNodeNames := make(map[string]string)
+		for _, pod := range res.Spec.PodsStatus {
+			if pod.PodIP == res.Spec.ClusterData.KeepersState[res.Spec.ClusterData.ClusterView.Master].ListenAddress {
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/upgrade.go]: I wonder - if you're referring to master so frequently - would it not be easier if `KeepersState` kept the master state separately? Which would also be mandatory - i.e. not a pointer

File: internal/stolonctl/pkg/cluster/upgrade.go [hunk 11]
```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
 
 	// Backup Postgres
-	if !crdclient.IsPhaseCompleted(res, crd.StolonUpgradePhaseBackupPostgres) {
-		res, err = crdclient.MarkPhase(res, crd.StolonUpgradePhaseBackupPostgres, crd.StolonUpgradeStatusInProgress)
-		if err != nil {
-			return trace.Wrap(err)
+	res, err = upgradeControl.executePhase(res, crd.StolonUpgradePhaseBackupPostgres, func() error {
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/upgrade.go]: The `Upgrade` function feels cluttered - I wonder if you could factor some things out - like move the phases to individual functions and execute them in a loop:
```go
for _, phase := range phases {
	result, err := upgradeControl.executePhase(result, phase.name, phase.handler(upgradeControl))
	// Handle error
}
```

File: images/jobs/keeper-upgrade.sh [hunk 12]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+# -*- mode: sh; -*-
+
+# File: entrypoint.sh
```
[the remark this answers is not in the document]
Comment (UnderGreen) [reply images/jobs/keeper-upgrade.sh]: Nope. It is snippet, which I should update to use Gravitational Inc instead of my name. :)

File: internal/stolonctl/pkg/cluster/upgrade.go (hunk 3 above)
Comment (UnderGreen) [reply internal/stolonctl/pkg/cluster/upgrade.go]: Actually I need rollback only for master node. Good catch, I keep old directory of postgresql. If schema upgrade fails here, I can move old directory to the right place and keep upgrade idempotent then.

File: internal/stolonctl/pkg/cluster/status.go [hunk 13]
```diff
@@ -106,25 +107,18 @@ type Status struct {
...
+func (s *Status) getMasterStatus() (*crd.MasterStatus, error) {
 	for _, pod := range s.PodsStatus {
-		if pod.PodIP == s.ClusterData.KeepersState[s.ClusterData.ClusterView.Master].ListenAddress {
-			return &MasterStatus{
-				PodName: pod.Name,
-				Healthy: s.ClusterData.KeepersState[s.ClusterData.ClusterView.Master].Healthy,
-				HostIP:  pod.HostIP,
-			}, nil
+		if _, ok := s.ClusterData.KeepersState[s.ClusterData.ClusterView.Master]; ok {
+			if pod.PodIP == s.ClusterData.KeepersState[s.ClusterData.ClusterView.Master].ListenAddress {
```
Comment (a-palchikov) [inline internal/stolonctl/pkg/cluster/status.go]: you could avoid double access by storing the result in a variable:
```go
if master, ok := s.ClusterData.KeepersState[s.ClusterData.ClusterView.Master]; ok {
			if pod.PodIP == master.ListenAddress {
```
I'd also invert the condition to check for value: `master != nil`
